File: src/types/user-import.ts
/**
 * Types for CSV/Excel user import functionality
 */

/**
 * Wizard step identifiers
 */
export type WizardStep = 'upload' | 'map' | 'review' | 'results'

/**
 * Single user row from CSV/Excel import
 */
export interface UserImportRow {
  username: string
  email: string
  firstname?: string
  lastname?: string
  password?: string
}

/**
 * Column mapping configuration
 */
export interface ColumnMapping {
  /** CSV column header */
  csvColumn: string
  /** Target field (username, email, firstname, lastname, password, or null for ignore) */
  targetField: 'username' | 'email' | 'firstname' | 'lastname' | 'password' | null
  /** Whether this field is required (username and email are required) */
  isRequired: boolean
}

/**
 * Validation error for a specific row/field
 */
export interface ValidationError {
  /** Row index (0-based) */
  row: number
  /** Field that has the error */
  field: string
  /** Error message */
  message: string
}

/**
 * Result for a single user import attempt
 */
export interface UserImportResult {
  /** Row index (0-based) */
  row_index: number
  /** Username that was attempted */
  username: string
  /** Email that was attempted */
  email: string
  /** Whether import was successful */
  success: boolean
  /** Error message if failed */
  error_message?: string
  /** Keycloak user ID if created successfully */
  user_id?: string
  /** Generated password if one was created */
  generated_password?: string
}

/**
 * Response from bulk import API
 */
export interface BulkImportResponse {
  /** Number of successfully imported users */
  success_count: number
  /** Number of failed imports */
  error_count: number
  /** Total number of users attempted */
  total_count: number
  /** Detailed results per row */
  results: UserImportResult[]
}

/**
 * Request body for bulk import API
 */
export interface BulkImportRequest {
  /** Target organization ID */
  organization_id: string
  /** Users to import */
  users: UserImportRow[]
  /** Whether to generate passwords for users without passwords */
  generate_passwords: boolean
}

/**
 * Parsed file data from CSV/Excel
 */
export interface ParsedFileData {
  /** Original file name */
  fileName: string
  /** File size in bytes */
  fileSize: number
  /** Column headers from the file */
  headers: string[]
  /** Data rows (array of arrays, each inner array is a row) */
  rows: string[][]
  /** Total row count (excluding header) */
  rowCount: number
}

/**
 * Duplicate detection result
 */
export interface DuplicateInfo {
  /** Email that is duplicated */
  email: string
  /** Username associated with the duplicate email */
  username: string
  /** Row index where duplicate was found */
  rowIndex: number
  /** Whether it's a duplicate within the file or against existing users */
  type: 'internal' | 'existing'
}

/**
 * Column pattern for auto-detection
 */
export interface ColumnPattern {
  /** Target field name */
  field: 'username' | 'email' | 'firstname' | 'lastname' | 'password'
  /** Patterns to match (lowercase) */
  patterns: string[]
}

/**
 * Column patterns for auto-detection
 * Based on spec requirements
 */
export const COLUMN_PATTERNS: ColumnPattern[] = [
  {
    field: 'username',
    patterns: ['username', 'user_name', 'login', 'user', 'utilisateur'],
  },
  {
    field: 'email',
    patterns: ['email', 'e-mail', 'mail', 'emails', 'courriel', 'e_mail'],
  },
  {
    field: 'firstname',
    patterns: ['firstname', 'first_name', 'prenom', 'given_name', 'first', 'prénom'],
  },
  {
    field: 'lastname',
    patterns: ['lastname', 'last_name', 'nom', 'family_name', 'last', 'name', 'surname'],
  },
  {
    field: 'password',
    patterns: ['password', 'pwd', 'pass', 'mot_de_passe'],
  },
]

/**
 * Maximum number of rows allowed per import
 */
export const MAX_IMPORT_ROWS = 100

/**
 * Supported file extensions
 */
export const SUPPORTED_FILE_EXTENSIONS = ['.csv', '.xlsx'] as const
export type SupportedFileExtension = (typeof SUPPORTED_FILE_EXTENSIONS)[number]

File: apps/screen/app/schemas/user_import.py
"""Schemas for bulk user import functionality."""

import re

from pydantic import BaseModel, EmailStr, field_validator


class UserImportRow(BaseModel):
    """Single user row from CSV/Excel import.

    All fields except username and email are optional.
    Password is optional - if not provided and generate_passwords is True,
    a random password will be generated.
    """

    username: str
    email: EmailStr
    firstname: str | None = None
    lastname: str | None = None
    password: str | None = None

    @field_validator("username")
    @classmethod
    def validate_username(cls, v: str) -> str:
        """Validate username format."""
        v = v.strip()
        if not v:
            raise ValueError("Username cannot be empty")
        if len(v) < 3:
            raise ValueError("Username must be at least 3 characters")
        if len(v) > 50:
            raise ValueError("Username must be 50 characters or less")
        # Keycloak username pattern: alphanumeric, dots, underscores, hyphens
        if not re.match(r"^[a-zA-Z0-9._-]+$", v):
            raise ValueError("Username can only contain letters, numbers, dots, underscores, and hyphens")
        return v.lower()  # Normalize to lowercase

    @field_validator("firstname", "lastname", mode="before")
    @classmethod
    def strip_whitespace(cls, v: str | None) -> str | None:
        """Strip whitespace from names."""
        if v is None:
            return None
        v = v.strip()
        return v if v else None

    @field_validator("password")
    @classmethod
    def validate_password(cls, v: str | None) -> str | None:
        """Validate password if provided."""
        if v is None or v == "":
            return None

        if len(v) < 8:
            raise ValueError("Password must be at least 8 characters")

        # Check complexity requirements
        has_upper = any(c.isupper() for c in v)
        has_lower = any(c.islower() for c in v)
        has_digit = any(c.isdigit() for c in v)
        has_special = any(c in "!@#$%^&*()_+-=[]{}|;:,.<>?" for c in v)

        if not (has_upper and has_lower and has_digit and has_special):
            raise ValueError("Password must contain uppercase, lowercase, number, and special character")

        return v


class BulkUserImportRequest(BaseModel):
    """Request body for bulk user import.

    Args:
        organization_id: Target Keycloak organization UUID
        users: List of user rows to import (max 100)
        generate_passwords: Whether to generate random passwords for users without passwords
    """

    organization_id: str
    users: list[UserImportRow]
    generate_passwords: bool = True

    @field_validator("organization_id")
    @classmethod
    def validate_organization_id(cls, v: str) -> str:
        """Validate organization_id is not empty."""
        v = v.strip()
        if not v:
            raise ValueError("Organization ID cannot be empty")
        return v

    @field_validator("users")
    @classmethod
    def validate_users_count(cls, v: list[UserImportRow]) -> list[UserImportRow]:
        """Validate user count limits."""
        if len(v) == 0:
            raise ValueError("At least one user is required")
        if len(v) > 100:
            raise ValueError("Maximum 100 users per import")
        return v


class UserImportResult(BaseModel):
    """Result for a single user import attempt.

    Contains success status and error message if failed.
    For successful imports with generated passwords, includes the password.
    """

    row_index: int  # 0-based index in the original import array
    username: str
    email: str
    success: bool
    error_message: str | None = None
    user_id: str | None = None  # Keycloak user UUID if created
    generated_password: str | None = None  # Only if password was generated


class BulkUserImportResponse(BaseModel):
    """Response for bulk user import operation.

    Provides summary counts and detailed results per row.
    Supports partial success - some rows may fail while others succeed.
    """

    success_count: int
    error_count: int
    total_count: int
    results: list[UserImportResult]


class DuplicateCheckResult(BaseModel):
    """Result of duplicate email/username check.

    Used internally before import to identify duplicates.
    """

    username: str
    email: str
    row_index: int
    is_duplicate_email: bool = False
    is_duplicate_username: bool = False
    existing_user_id: str | None = None
